### PR TITLE
expression: support ilike function

### LINKED ISSUE #40943: Support ilike function
## Feature Request

**Is your feature request related to a problem? Please describe:**

**Describe the feature you'd like:**

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by xzhangxian1008**
### What problem does this PR solve?

Issue Number: close #40943

Description: We copy most of the `like` function codes to implement the ilike function.

### What is changed and how it works?

- Move some functions to `builtin_func_param.go`
- Modify the parser to support `ilike`
- Change the name of regexpParam to funcParam
- support `ilike` function

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: expression/builtin_ilike.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+		return nil, err
+	}
+	argTp := []types.EvalType{types.ETString, types.ETString, types.ETInt}
+	bf, err := newBaseBuiltinFuncWithTp(ctx, c.funcName, args, types.ETInt, argTp...)
+	if err != nil {
+		return nil, err
+	}
+	bf.tp.SetFlen(1)
+	sig := &builtinIlikeSig{bf, nil, false, sync.Once{}}
+	sig.setPbCode(tipb.ScalarFuncSig_LikeSig) // TODO change it
```
Comment (windtalker) [inline expression/builtin_ilike.go:51]: I think it need to be updated in this pr?

File: expression/builtin_ilike.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+	valStr = stringutil.LowerOneString(valStr)
+	if stringutil.IsUpperAscii(rune(escape)) {
+		patternStr = stringutil.LowerOneStringExcludingSpecificChar(patternStr, rune(escape))
+	} else {
+		patternStr = stringutil.LowerOneString(patternStr)
+	}
+
+	memorization := func() {
+		if b.pattern == nil {
+			b.pattern = collate.ConvertAndGetBinCollation(b.collation).Pattern()
```
Comment (windtalker) [inline expression/builtin_ilike.go:98]: even if the orignal collation is ci based, here still use bin collation?

Comment (xzhangxian1008) [reply expression/builtin_ilike.go:98]: > even if the orignal collation is ci based, here still use bin collation?

yes, or the behaviour will not be consistent for ilike in different collation.

File: expression/builtin_ilike.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+	}
+
+	escape, isNull, err := b.args[2].EvalInt(b.ctx, row)
+	if isNull || err != nil {
+		return 0, isNull, err
+	}
+
+	valStr = stringutil.LowerOneString(valStr)
+	if stringutil.IsUpperAscii(rune(escape)) {
+		patternStr = stringutil.LowerOneStringExcludingSpecificChar(patternStr, rune(escape))
```
Comment (windtalker) [inline expression/builtin_ilike.go:91]: I think it's not enough, consider `string = 'ABC'`, `escape char = 'a'`, then `A` will be escaped

Comment (xzhangxian1008) [reply expression/builtin_ilike.go]: > I think it's not enough, consider `string = 'ABC'`, `escape char = 'a'`, then `A` will be escaped

I will fix it

File: expression/builtin_ilike.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+	escape, isNull, err := b.args[2].EvalInt(b.ctx, row)
+	if isNull || err != nil {
+		return 0, isNull, err
+	}
+
+	valStrBytes := []byte(valStr)
+	patternStrBytes := []byte(patternStr)
+
+	stringutil.LowerOneString(valStrBytes)
```
Comment (windtalker) [inline expression/builtin_ilike.go:92]: will this modify the `valStr` inplace?

File: expression/builtin_ilike_vec.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+	}
+	escapes := bufEscape.Int64s()
+
+	// Must not use b.pattern to avoid data race
+	pattern := collate.ConvertAndGetBinCollation(b.collation).Pattern()
+
+	tmpValCol := bufVal.CopyConstruct(nil)
+	tmpPatternCol := bufPattern.CopyConstruct(nil)
+	LowerAlphaAscii(tmpValCol, rowNum)
+	LowerAlphaAsciiExcludeEscapeChar(tmpPatternCol, rowNum, escapes)
```
Comment (windtalker) [inline expression/builtin_ilike_vec.go:82]: will `escape` column be modified inplace?

File: expression/builtin_ilike.go (hunk 4 above)
Comment (xzhangxian1008) [reply expression/builtin_ilike.go:92]: no

File: expression/builtin_ilike_vec.go (hunk 5 above)
Comment (xzhangxian1008) [reply expression/builtin_ilike_vec.go:82]: > will `escape` column be modified inplace?

yes, and i will fix it

File: util/stringutil/string_util.go [hunk 6]
```diff
@@ -406,3 +406,79 @@ func ConvertPosInUtf8(str *string, pos int64) int64 {
...
+		if IsUpperASCII(str[i]) {
+			str[i] = toLowerIfAlphaASCII(str[i])
+		}
+	}
+}
+
+// LowerOneStringExcludeEscapeChar, Sometimes we want to lower strings and exclude an escape char
+//
+// When escape_char is a lower char, we need to convert it to the capital char
+// Because: when lowering "ABC" with escape 'a', after lower, "ABC" -> "abc",
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply util/stringutil/string_util.go:447]: > Maybe add more examples to show the convert logic.

okk

File: expression/builtin_ilike_vec.go [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+	result.ResizeInt64(rowNum, false)
+	result.MergeNulls(bufVal, bufPattern, bufEscape)
+	i64s := result.Int64s()
+
+	for i := 0; i < rowNum; i++ {
+		if result.IsNull(i) {
+			continue
+		}
+		pattern.Compile(bufPattern.GetString(i), byte(escapes[i]))
```
Comment (windtalker) [inline expression/builtin_ilike_vec.go:97]: Why don't check if it can memorize first ?

File: expression/builtin_ilike_vec.go [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+	if params[0].getCol() == nil {
+		return b.constVec(params[0].getStringVal(0), params[1], rowNum, escape, result)
+	}
+
+	return b.vecVec(params, rowNum, escape, result)
+}
+
+func (b *builtinIlikeSig) vecEvalInt(input *chunk.Chunk, result *chunk.Column) error {
+	rowNum := input.NumRows()
+	params := make([]*regexpParam, 0, 3)
```
Comment (windtalker) [inline expression/builtin_ilike_vec.go:174]: Looks like `regexpParam` is no longer a suitable name, should rename it.

Comment (xzhangxian1008) [reply expression/builtin_ilike_vec.go:174]: > Looks like `regexpParam` is no longer a suitable name, should rename it.

okk

**Review by windtalker — APPROVED** [review]
LGTM

File: expression/builtin_ilike_vec.go [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+
+import (
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/collate"
+	"github.com/pingcap/tidb/util/hack"
+	"github.com/pingcap/tidb/util/stringutil"
+)
+
+func LowerAlphaAscii(lowered_col *chunk.Column, rowNum int) {
```
Comment (XuHuaiyu) [inline expression/builtin_ilike_vec.go:26]: loweredCol

File: expression/builtin_ilike_vec.go [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+			continue
+		}
+		b.pattern.Compile(param.getStringVal(i), byte(escape))
+		match := b.pattern.DoMatch(expr)
+		i64s[i] = boolToInt64(match)
+	}
+	return nil
+}
+
+func (b *builtinIlikeSig) ilikeWithMemorization(exprParam *funcParam, rowNum int, escape int64, result *chunk.Column) error {
```
Comment (XuHuaiyu) [inline expression/builtin_ilike_vec.go:152]: useless input parameter escape

Comment (xzhangxian1008) [reply expression/builtin_ilike_vec.go:152]: I will delete it

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**